BugFix for function sparse_coo_tensor_ctor

aocsa opened the pull request:
This PR solves a bug when the user does not specify the device for the sparse tensor using the `sparse_coo_tensor` function but the values and indices tensor are in different CUDA device than default. 

Ex:
`sparse_tensor = torch.sparse_coo_tensor(sp_ind, sp_val, sp_shape)`.

This was solved by using the same device_index from values Tensor for  indices Tensor at `sparse_coo_tensor_ctor` function 

Resolves bug #[28500](https://github.com/pytorch/pytorch/issues/28500).

Note that `torch.sparse.FloatTensor` constructor worked as expected because it uses  the `legacy_sparse_tensor_ctor`.  This function uses the same device information of the original Tensor PyThon object. In this fix we replicated this behavior in the new   constructor `sparse_coo_tensor_ctor` function. 

So now this works!
```
import torch
sp_val = torch.rand(10).to('cuda:1')
print(sp_val, sp_val.device)

sp_ind = torch.tensor([[0,1,2,3,4,5,6,7,8,9], [0,1,2,3,4,5,6,7,8,9]], device='cuda:1')
print(sp_ind)

x = torch.sparse_coo_tensor(sp_ind, sp_val, (10, 10))
```

pearu reviewed (commented):
The bug fix looks good to me.

@aocsa, could you implement also the corresponding unittest?

aocsa commented:
> The bug fix looks good to me.
> 
> @aocsa, could you implement also the corresponding unittest?

Thanks @pearu, I update this PR adding the corresponding unittest when the number of available devices is greater than 1.

pearu reviewed (approved):
LGTM.
The CI failures seem to be unrelated to the changes in this PR.

ngimel commented:
So behavior introduced by this PR (unless I'm misreading) is:
1) if device is specified, the sparse tensor will be created on this device.  
2) if device is unspecified, and dtype is also unspecified, the sparse tensor will be created with dtype and device of vals. 
3) if device is unspecified, but dtype is specified, the sparse tensor will be created on default device, and device of val will be ignored. Why difference with the previous case? Is it documented somewhere?

ngimel commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -236,7 +236,8 @@ Tensor internal_new_from_data(
     bool type_inference,
-    bool pin_memory = false) {
+    bool pin_memory = false,
+    c10::DeviceIndex device_index = -1) {
```
having both `device_index` and `device_opt` arguments with unclear precedence, and computing actual device based not only on these 2 arguments, but also on `type_inference`, is super confusing.

aocsa replied:
Thanks for the feedback @ngimel.   

> So behavior introduced by this PR (unless I'm misreading) is:
> 
> 1. if device is specified, the sparse tensor will be created on this device.
Yes, It will.  Nothing changed here.

> 2. if device is unspecified, and dtype is also unspecified, the sparse tensor will be created with dtype and device of vals.
Yes, It will. Basically, this fix cover this scenario by using the same device information of the vals tensor including the `device_index` for the creation of the indices tensor. Before this change, the function `internal_new_from_data` assumed that the vals tensor was by default at 'cuda:0' and it failed when values tensor was set up at 'cuda:1' for example.  

> 3. if device is unspecified, but dtype is specified, the sparse tensor will be created on default device, and device of val will be ignored. Why difference with the previous case? Is it documented somewhere?
Yes the device info will be ignored and will use by default 'cpu' in this case the `device_index` is optional and unnecessary.   

>> having both device_index and device_opt arguments with unclear precedence, and computing actual device based not only on these 2 arguments, but also on type_inference, is super confusing.

I agree with that. Let me refactor this just a little. The Parameter should be a tuple of  <c10::DispatchKey, c10::DeviceIndex> to identify the default `dispatch_key` device. So, in case that vals tensor is set up for example at device 'cuda:1', the indices tensor should be in the same device.

ngimel replied:
Let's step back and spell out expected behavior though. I propose if device or dtype is unspecified, fill it in from vals, which is not what this PR does. So
1) device and dtype are specified - use them
2) device and dtype are unspecified - fill in from vals
3) device is specified, dtype is unspecified - use device, fill in dtype from vals
4) device is unspecified, dtype is specified - fill in device from vals, use dtype
Does it make sense?

mruberry replied:
This seems OK, but for safety what if the function throws an error when the device is unspecified and the indices and value tensors have a device mismatch? So a slight change to:

1) if device specified, use it; if dtype is specified, use it
2) if device is unspecified, check that both indices and values are on the same device, if they are then use that device, if they're not then throw a Runtime error ("Expected indices and values to be on the same device, but got ..."
3) if dtype is unspecified, use the dtype of the values tensor

aocsa replied:
> Let's step back and spell out expected behavior though. I propose if device or dtype is unspecified, fill it in from vals, which is not what this PR does. So
> 
> 1. device and dtype are specified - use them
All right. No changes required for this case.

> 2. device and dtype are unspecified - fill in from vals
All right. This fix cover this case. 

> 3. device is specified, dtype is unspecified - use device, fill in dtype from vals
All right. No changes required for this case.

> 4. device is unspecified, dtype is specified - fill in device from vals, use dtype
All right. No changes required for this case.

>    Does it make sense?

aocsa replied:
> This seems OK, but for safety what if the function throws an error when the device is unspecified and the indices and value tensors have a device mismatch? So a slight change to:

That is a good question. Inside the function `sparse_coo_tensor_ctor`  function `internal_new_from_data` is called twice once for values and then one for indices tensors. If there is a mismatch the function `internal_new_from_data`  will assure that both (values and indices) are going to be in the same device by calling `return var.to(device, inferred_scalar_type, ...)`/ 

> 
> 1. if device specified, use it; if dtype is specified, use it

All right. No changes required for this case.

> 2. if device is unspecified, check that both indices and values are on the same device, if they are then use that device, if they're not then throw a Runtime error ("Expected indices and values to be on the same device, but got ..."

I have my concerns here. If I add this exception.  The code related to make sure that both values and indices are using the same device though the function  `return var.to(device, inferred_scalar_type, ...)` will be unnecessary, and we will lose this functionality. Moreover, some unit tests rely on this functionality.  

```
https://pytorch.org/docs/master/generated/torch.sparse_coo_tensor.html
device (torch.device, optional) – the desired device of returned tensor. Default: if None, uses the current device for the default tensor type (see torch.set_default_tensor_type()). device will be the CPU for CPU tensor types and the current CUDA device for CUDA tensor types.
```
So, maybe it is better to clarify in the TORCH.SPARSE_COO_TENSOR documentation the  expected behavior for this case. 

> 3. if dtype is unspecified, use the dtype of the values tensor
All right. No changes required for this case.

Looking forward your comments, cc @mruberry

ngimel replied:
Re: the 4th case, my proposal differs from what's currently implemented, that's true. But IMO it makes it more consistent, in that missing data is always filled from the vals tensor. Otherwise, missing data is filled from the vals everywhere (cases 2 and 3) except when only device is missing, and then default device is used.

aocsa replied:
Thanks for the feedback @ngimel. So just to be clear.  If device is unspecified, dtype is specified - fill in device from vals, use dtype. However, if device is unspecified and indices and values are not on the same device,  I suggest just printing a warning ("Expected indices and values to be on the same device, but got ..."   instead of a runtime error in order to maintain the same expected behavior as `torch.sparse.FloatTensor` constructor.  Also,  we will use the vals' device as a default device for indices. 

Looking forward your comments.

ngimel replied:
Yes, this seems fine, warn if vals and indices are on the different devices, and create vals and indices on the vals device.

aocsa commented:
> So behavior introduced by this PR (unless I'm misreading) is:
> 
> 1. if device is specified, the sparse tensor will be created on this device.
> 2. if device is unspecified, and dtype is also unspecified, the sparse tensor will be created with dtype and device of vals.
> 3. if device is unspecified, but dtype is specified, the sparse tensor will be created on default device, and device of val will be ignored. Why difference with the previous case? Is it documented somewhere?

mruberry commented:
Sorry for arriving late. @aocsa, did you mean to close this PR? It would be nice to fix this bug. Is there something we can help with?

aocsa commented:
> Sorry for arriving late. @aocsa, did you mean to close this PR? It would be nice to fix this bug. Is there something we can help with?

Ohh  sorry @mruberry. I accidentally press the wrong button. I am still working in this issue.

mruberry commented:
> > Sorry for arriving late. @aocsa, did you mean to close this PR? It would be nice to fix this bug. Is there something we can help with?
> 
> Ohh sorry @mruberry. I accidentally press the wrong button. I am still working in this issue.

No worries. I'll take a look now, then.

mruberry commented:
> I have my concerns here. If I add this exception. The code related to make sure that both values and indices are using the same device though the function return var.to(device, inferred_scalar_type, ...) will be unnecessary, and we will lose this functionality. Moreover, some unit tests rely on this functionality.
> 
> So, maybe it is better to clarify it in the TORCH.SPARSE_COO_TENSOR documentation the expected behavior for this case.

OK.

aocsa commented:
> > I have my concerns here. If I add this exception. The code related to make sure that both values and indices are using the same device though the function return var.to(device, inferred_scalar_type, ...) will be unnecessary, and we will lose this functionality. Moreover, some unit tests rely on this functionality.
> > So, maybe it is better to clarify it in the TORCH.SPARSE_COO_TENSOR documentation the expected behavior for this case.
> 
> OK.

Quick update: This PR was updated based on latest feedback. cc @mruberry

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5646,7 +5646,9 @@ def merge_dicts(*dicts):
     device (:class:`torch.device`, optional): the desired device of returned tensor.
         Default: if None, uses the current device for the default tensor type
```
This documentation is outdated, since we don't have a notion of "current device for the default tensor type" anymore. Plus, with your change, when device is None then PyTorch will produce a sparse tensor on the same device as the values tensor.

mruberry replied:
(The above note is outdated based on offline discussion.)

aocsa replied:
Thanks, Mike! I updated the PR based on latest feedback. cc @mruberry

mruberry commented:
Logic looks good but looks like you need to merge with master or rebase (to help the CI merge the branch) and fix the flake8 and clang-tidy lint issues. Once that's done we should be good to go!

aocsa committed:
minor change: solves magic number warning

aocsa commented:
> Logic looks good but looks like you need to merge with master or rebase (to help the CI merge the branch) and fix the flake8 and clang-tidy lint issues. Once that's done we should be good to go!

I rebased this PR and fixed the flake8 and clang-tidy lint issues.  
Thanks @mruberry

ngimel commented:
This still results in unexpected behavior if dtype is specified, but device is not (because in this case type_inference is false):
```
In [18]: i=torch.tensor(([0], [2]), device="cuda:1", dtype=torch.long)                                                                                                                                
In [19]: v=torch.tensor([1.], device="cuda:1")                                                                                                                                                        
In [20]: t=torch.sparse_coo_tensor(i, v, dtype=torch.float64)                                                                                                                                         
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-20-09518dc143ad> in <module>
----> 1 t=torch.sparse_coo_tensor(i, v, dtype=torch.float64)
RuntimeError: Expected indices and values to be on the same device, but got: values cpu, indices: cuda:1
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5710,9 +5710,7 @@ def merge_dicts(*dicts):
-        Default: if None, uses the current device for the default tensor type
-        (see :func:`torch.set_default_tensor_type`). :attr:`device` will be the CPU
-        for CPU tensor types and the current CUDA device for CUDA tensor types.
+        Default: if None, the device of indices must match device of values, otherwise an exception is raised. 
```
More correctly it would be "and the sparse tensor is constructed on the same device?"

ngimel replied:
Both statements are technically correct.

aocsa commented:
> This still results in unexpected behavior if dtype is specified, but device is not (because in this case type_inference is false):
> 
> ```
> In [18]: i=torch.tensor(([0], [2]), device="cuda:1", dtype=torch.long)                                                                                                                                
> In [19]: v=torch.tensor([1.], device="cuda:1")                                                                                                                                                        
> In [20]: t=torch.sparse_coo_tensor(i, v, dtype=torch.float64)                                                                                                                                         
> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-20-09518dc143ad> in <module>
> ----> 1 t=torch.sparse_coo_tensor(i, v, dtype=torch.float64)
> RuntimeError: Expected indices and values to be on the same device, but got: values cpu, indices: cuda:1
> ```

Thanks for notice me about this issue @ngimel. For this case (dtype is specified, but device is not), in the current master branch, no matter where is the indices or the values both are going to be transformed to default  device (CPU).    See example below. 

```
import torch
i=torch.tensor(([0], [2]), device="cuda:0", dtype=torch.long)
v=torch.tensor([1.], device="cuda:1")           
t=torch.sparse_coo_tensor(i, v, dtype=torch.float64)    
print(t.device) # CPU
```

Moreover, If I understood well we should avoid internal transformation of the device source. So, I am not sure,  what should be the expected behavior for this case.

ngimel commented:
Existing behavior is very strange. The expected behavior that we should document is when device is not specified (regardless of whether dtype is specified or not) indices and values should be on the same device, and the sparse tensor will be constructed on the this device.

mruberry commented on test/test_sparse.py:
```diff
@@ -1746,6 +1746,16 @@ def check_device(x, device_id):
+        i = self.index_tensor([[2]]).cuda(1)
+        v = self.value_empty(1, 0).cuda(1)
+        x = torch.sparse_coo_tensor(i, v, torch.Size([3, 0]))
+        check_device(x, 1)
```
Better add a test where dtype is set here, too

aocsa replied:
all right!

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -251,7 +251,7 @@ Tensor internal_new_from_data(
     // are defined per-layout-type (e.g. tensor vs sparse_coo_tensor).
     const auto& inferred_scalar_type = type_inference ? var.scalar_type() : scalar_type;
-    auto device = device_opt.has_value() ? *device_opt : (type_inference ? var.device() : at::Device(computeDeviceType(dispatch_key)));
+    auto device = device_opt.has_value() ? *device_opt : (type_inference ? var.device() : at::Device(computeDeviceType(dispatch_key.first), dispatch_key.second));
```
I think changes like this can be reverted?

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -262,7 +262,7 @@ Tensor internal_new_from_data(
     auto tensor = tensor_from_cuda_array_interface(data);
     const auto& inferred_scalar_type = type_inference ? tensor.scalar_type() : scalar_type;
-    auto device = device_opt.has_value() ? *device_opt : at::Device(computeDeviceType(dispatch_key));
+    auto device = device_opt.has_value() ? *device_opt : at::Device(computeDeviceType(dispatch_key.first));
```
Why in this case and the following case does the call not also include the index?

aocsa replied:
Because this is the case when the tensor comes from `numpy`, so, the index is not needed.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -605,27 +642,55 @@ Tensor sparse_coo_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
-    Tensor indices = internal_new_from_data(legacyExtractDispatchKey(values.key_set()), kLong, r.deviceOptional(3), r.pyobject(0), false, true, false);
-    return at::sparse_coo_tensor(indices, values, values.options().layout(at::kSparse)).set_requires_grad(r.toBool(4));
+    Tensor values = internal_new_from_data(inferred_dispatch_key, inferred_scalar_type, r.deviceOptional(ARG_DEVICE), r.pyobject(ARG_VALUES), false, true, type_inference);
+    check_expected_devices(device_guard.original_device(), values, r.pyobject(ARG_INDICES));
```
What about moving this check to after the construction of the indices tensor and rewriting it as `TORCH_CHECK(values.device() == indices.device(), ...)'?

aocsa replied:
All right!

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -605,27 +642,55 @@ Tensor sparse_coo_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
+    const auto inferred_scalar_type = r.scalartypeWithDefault(ARG_TYPE, scalar_type);
+    at::OptionalDeviceGuard device_guard(r.deviceOptional(ARG_DEVICE));
+    Tensor values = internal_new_from_data(inferred_dispatch_key, inferred_scalar_type, r.deviceOptional(ARG_DEVICE), r.pyobject(ARG_VALUES), false, true, type_inference);
+    check_expected_devices(device_guard.original_device(), values, r.pyobject(ARG_INDICES));
```
Same question as above about the position and style of this check

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -651,7 +716,9 @@ Tensor _sparse_coo_tensor_unsafe_ctor(c10::DispatchKey dispatch_key, at::ScalarT
   at::OptionalDeviceGuard device_guard(r.deviceOptional(ARG_DEVICE));
   Tensor values = internal_new_from_data(inferred_dispatch_key, inferred_scalar_type, r.deviceOptional(ARG_DEVICE), r.pyobject(ARG_VALUES), false, true, type_inference);
-  Tensor indices = internal_new_from_data(legacyExtractDispatchKey(values.key_set()), kLong, r.deviceOptional(ARG_DEVICE), r.pyobject(ARG_INDICES), false, true, false);
+  check_expected_devices(device_guard.original_device(), values, r.pyobject(0));
```
Same question as above

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -251,7 +251,7 @@ Tensor internal_new_from_data(
     // are defined per-layout-type (e.g. tensor vs sparse_coo_tensor).
     const auto& inferred_scalar_type = type_inference ? var.scalar_type() : scalar_type;
-    auto device = device_opt.has_value() ? *device_opt : (type_inference ? var.device() : at::Device(computeDeviceType(dispatch_key)));
+    auto device = device_opt.has_value() ? *device_opt : (type_inference ? var.device() : at::Device(computeDeviceType(dispatch_key.first), dispatch_key.second));
```
This is the existing line that needs to change to no longer cause dtype to have an affect on device, right? We really don't want there to be any connection between dtype and device.

aocsa replied:
Sure, but it can be solved by using the right `dispatch_key`.  For now, I will not touch this `internal_new_from_data`  function.

mruberry commented:
@aocsa, I had a chance to review/investigate the dtype issue more with @ngimel, and we actually found that it affects other functions, like `torch.as_tensor`, too, in addition to being a source of confusion for how best to address this sparse ctor issue. Sorry for that existing issue getting conflated with what you're trying to fix.

To simplify this fix and address that underlying issue, we thought we'd fix `internal_new_from_data`'s behavior. That should separate these issues and make this much easier to address. 

What are your thoughts with that approach?

aocsa commented:
> @aocsa, I had a chance to review/investigate the dtype issue more with @ngimel, and we actually found that it affects other functions, like `torch.as_tensor`, too, in addition to being a source of confusion for how best to address this sparse ctor issue. Sorry for that existing issue getting conflated with what you're trying to fix.
> 
> To simplify this fix and address that underlying issue, we thought we'd fix `internal_new_from_data`'s behavior. That should separate these issues and make this much easier to address.
> 
> What are your thoughts with that approach?

Sure, I think it's a good idea. 
Actually,  that was my initial approach, don't touch the internal mechanism of `internal_new_from_data` function due to it is used in other parts. 

I am already testing new changes that only touch the `sparse_coo_tensor_ctor` function, choosing by default the `dispatch_key` device info from values tensor. I can update my PR covering the latest feedback and then wait for `internal_new_from_data`fix. 

cc @mruberry

mruberry commented:
Great; I'll let you know when the fix is available.

aocsa committed:
minor update covering dtype issue

mruberry commented:
Fyi https://github.com/pytorch/pytorch/pull/41984, which is currently being tested to see if it breaks any existing PyTorch programs we know of.

mruberry commented:
#41984 is in!

@aocsa, an unintended side affect of that change is that it may also have fixed this spare issue. Would you please review? It now appears like sparse tensors do have to be on the same device if the device kwarg isn't filled out.

aocsa commented:
> #41984 is in!
> 
> @aocsa, an unintended side affect of that change is that it may also have fixed this spare issue. Would you please review? It now appears like sparse tensors do have to be on the same device if the device kwarg isn't filled out.

Great!, I already reviewed and tested changes in PR #41984 with the identified test cases and it solves this issue too. So this PR should be closed, right? 
cc @mruberry

mruberry commented:
> > #41984 is in!
> > @aocsa, an unintended side affect of that change is that it may also have fixed this spare issue. Would you please review? It now appears like sparse tensors do have to be on the same device if the device kwarg isn't filled out.
> 
> Great!, I already reviewed and tested changes in PR #41984 with the identified test cases and it solves this issue too. So this PR should be closed, right?
> cc @mruberry

If you've verified the behavior then I think we're OK closing this, yes. I was surprised the change we identified as necessary here also solved this issue. Thank you for helping with it.
